### ISSUE TITLE
System.Text.Json code gen: unhelpful error when custom serializer context is not decorated with any JsonSerializable

### ISSUE THREAD

**Issue by Jericho** [dotnet/runtime#58698]
When my custom serialiazer context class has zero JsonSerializable attributes, I get an error message which does not help me understand what the problem is.

I simply get the dreaded:
```
CS0534	'MyJsonSerializerContext' does not implement inherited abstract member 'JsonSerializerContext.GetTypeInfo(Type)'
```

I do not get any additional information to help me realize that I forgot to add JsonSerializable attributes to my custom class. It would be extremely helpful to get an error message similar to `Your custom serializer context class must have at least one JsonSerializable attribute`.

Repro:

```csharp
public static class Program
{
    public static void Main()
    {
        Console.WriteLine("Hello World!");
    }
}

// Notice there are no JsonSerializable attributes here.
internal partial class MyJsonSerializerContext : JsonSerializerContext
{
}
```

As an aside: Is there a way to get more detailed information when something goes wrong with the code generator? Does it generate a log somewhere?

**Comment by Jericho** [comment]
By the way, I want to be completely clear that I fully understand that a custom serializer context without any JsonSerializable is not a valid scenario. I am not saying that the code generator should accept this situation. However, I think that a better/more descriptive error message would be helpful and would allow developers to resolve the situation easily.

**Comment by layomia** [comment]
Yes we can log a diagnostic error for this scenario.

> As an aside: Is there a way to get more detailed information when something goes wrong with the code generator? Does it generate a log somewhere?

Yes we log diagnostic errors for other scenarios, e.g. when a custom context class or one of its enclosing types is not `partial`.

**Comment by Jericho** [comment]
> Yes we log diagnostic errors for other scenarios

How do I access this log?

**Comment by layomia** [comment]
It's part of the error log shown during the build (visible both in command line & VS), just like the `CS0534` error you're seeing above.

**Comment by Jericho** [comment]
@layomia The reason I asked about a more detailed log is because `CS0534` is a very generic message that, unfortunately, does not give us any clue about what caused the problem. Consider this: let's say I was to raise an issue and said "I'm getting CS0534, help me figure out what's going on", I'm sure you would tell me that I haven't provided enough information. 

So, let me rephrase my question: how can we find more detailed information when we get generic errors like CS0534?

**Comment by layomia** [comment]
When you get `CS0534` when using the source generator, it means that there's another underlying issue that caused the generator not to emit the expected code. The generator emits various diagnostics which explain failure reasons, but it doesn't do it in all cases as this issue shows. We just need to emit a diagnostic for this scenario like proposed in https://github.com/dotnet/runtime/pull/58768.

A fix for this doesn't meet the bar to be ported to 6.0, so I'll move this issue to future.

**Comment by layomia** [comment]
Up for grabs. https://github.com/dotnet/runtime/pull/58768 attempted to fix this for 6.0 but we ran out of time. Note there'll be some [perf testing](https://github.com/dotnet/runtime/pull/58768#discussion_r704474361) to make sure we don't regress typing performance for large VS solutions.

**Comment by eiriktsarpalis** [comment]
We also need to make sure this diagnostic doesn't trigger for classes that provide an explicit override for the `GetTypeInfo` method. No doubt many users do this so this could trigger a source breaking change if we don't account for that case.

**Comment by eiriktsarpalis** [comment]
Moving to future for now as we figure out a way to do this without compromising editor performance.
